Azure:  Event support

bronaghs opened the pull request:
## Event support for Azure. 

### A couple of points to note:

* There is no payload included in the Azure event making it difficult to gather details.

* The Azure "write" event is fired after any configuration change has been made to a VM e.g. a change of flavor assignment, the addition of a new data disk etc, but it also is fired upon the creation of a new VM. Given the lack of payload and absence of any description, it is impossible to determine which scenario triggered the event.  

Corresponding Trello card:
https://trello.com/c/QnXCQPZj/50-3-azure-events-event-mappings

durandom commented on app/models/manageiq/providers/azure/cloud_manager/event_catcher/stream.rb:
```diff
@@ -0,0 +1,62 @@
+    Azure::Armrest::Insights::EventService.new(conf)
+  end
+
+  def each
```
I think `each` is not used anywhere. I removed it from the aws eventstream :)

durandom commented on app/models/manageiq/providers/azure/cloud_manager/event_catcher/stream.rb:
```diff
@@ -0,0 +1,62 @@
+class ManageIQ::Providers::Azure::CloudManager::EventCatcher::Stream
+  include Vmdb::Logging
```
is this from debugging? You dont seem to use logging in this class

durandom commented on app/models/manageiq/providers/azure/cloud_manager/event_catcher/stream.rb:
```diff
@@ -0,0 +1,62 @@
+
+      # HACK: the Azure Insights API does not support the 'gt' (greater than relational operator)
+      # therefore we have to poll from 1 second past the timestamp of the last event to avoid
+      # gathering the same event more than once.
```
I see you are fighting an ugly API here :(
But could this algorithm miss Events? Does this filter return _every_ event for "ge(since)" ? or is it paged and returns only the first 20?

bronaghs replied:
@durandom  - indeed!
So we are at risk of missing events that are fired between the timestamp of the last event and that timestamp + 1 second. I am in dialog with Microsoft on how they expect event consumers to work with this.
In addition, I double checked if events are returned in pages; apparently there is a "nextLink" property that is not present is ANY events I get back from Azure, even the last event in a poll. I am in dialog with Microsoft on this too. Depending on what conclusions we draw, I may have to update the code depending on their response. Thanks for raising this and having me visit this piece again.

chessbyte commented:
@blomquisg @Fryguy please review

bronaghs commented:
@blomquisg @Fryguy - ready for review

durandom commented on app/models/manageiq/providers/azure/cloud_manager/event_catcher/stream.rb:
```diff
@@ -0,0 +1,64 @@
+    while @collecting_events
+
+      # Grab only events for the last minute if this is the first poll
+      startup_interval = (Time.current - 1.minute).httpdate
```
this is only used for the first iteration of the loop.
how about inlining it or extracting a method?

you might even consider extracting the whole filter logic into their own methods and leave `each_batch` only concerned with yielding events

bronaghs replied:
@durandom - Yea, doing this made the code that bit more readable I think.

chessbyte commented on app/models/manageiq/providers/azure/event_catcher_mixin.rb:
```diff
@@ -0,0 +1,31 @@
+  end
+
+  def parse_vm_ref(event)
+    resource_id = event["resourceId"].downcase.split("/")
```
Can you make a comment with an example of what this looks like.  The use of `resource_id[2]`, `resource_id[4]`, `resource_id[6]`, and `resource_id[7]` leave an outsider very confused about what is being joined.

bronaghs replied:
@chessbyte  - Sure. Done.

blomquisg commented on app/models/manageiq/providers/azure/cloud_manager/event_catcher/stream.rb:
```diff
@@ -0,0 +1,70 @@
+    Time.zone.parse(events.last.event_timestamp).to_f + 0.001
+  end
+
+  def print_timestamp(time)
```
nitpick:  can you rename this method to `format_timestamp`?  When I read it, I thought it was going to output the timestamp somewhere. :)

blomquisg commented on app/models/manageiq/providers/azure/event_catcher_mixin.rb:
```diff
@@ -0,0 +1,38 @@
+    event_name = event["eventName"]["value"]
+    event_type = ""
+
+    unless event["authorization"].nil? || event["authorization"]["action"].nil?
```
I think you can skip these `nil` checks and then have `parse_event_action` handle it (see below).

Fryguy replied:
The latter is better, because parse_event_action has to fetch that path again...might as well only do it once in the parse_event_action method.

blomquisg commented:
A couple nitpicks.  Otherwise, LGTM

@Fryguy can you look at the config updates?

@gmcculloug can you look at the event handling updates?

gmcculloug commented:
@lfu Please review the event handling and suggest policy event actions where needed.

Fryguy commented on app/models/manageiq/providers/azure/cloud_manager/event_catcher/runner.rb:
```diff
@@ -0,0 +1,57 @@
+    # We do not care for 'Begin' events, only the 'End' events which indicate
+    # the action the event relates to was completed successfully.
+    event_type = parse_event_type(event)
+    filtered_events.include?(event_type) || event_type.end_with?("BeginRequest")
```
I'm not sure why we wouldn't collect Begin events...in other providers we try to collect everything.

bronaghs replied:
@Fryguy - I thought an *End* event was suffice and the *Begin* event was almost considered redundant. I thought this made sense given that the volume of events would double if the *Begin* events are stored. Do we keep Begin events for other providers?

Fryguy commented on app/models/manageiq/providers/azure/cloud_manager/event_catcher/stream.rb:
```diff
@@ -0,0 +1,70 @@
+    (Time.current - 1.minute).httpdate
+  end
+
+  def one_ms_from_last_timestamp(events)
```
Can you be sure to mark which methods are public and which should be private in this class and the other classes?

Fryguy commented on app/models/manageiq/providers/azure/cloud_manager/event_parser.rb:
```diff
@@ -0,0 +1,18 @@
+module ManageIQ::Providers::Azure::CloudManager::EventParser
+  extend ManageIQ::Providers::Azure::EventCatcherMixin
+
+  def self.event_to_hash(event, ems_id)
```
@blomquisg Is this supposed to be at the class level?  Or should we instantiate a parser?

durandom replied:
@Fryguy this is the way how all the other `EventParsers` do it. So I'd continue doing this.
But I have a rework of this cooking in #7104

Fryguy commented on app/models/manageiq/providers/azure/cloud_manager/event_parser.rb:
```diff
@@ -0,0 +1,18 @@
+      :message    => event["description"].blank? ? nil : event["description"],
+      :ems_id     => ems_id,
+      :event_type => event_type,
+      :full_data  => {:vm => {:uid_ems => parse_vm_ref(event)}},
```
I guess we only collect VM data?

Fryguy commented on app/models/manageiq/providers/azure/event_catcher_mixin.rb:
```diff
@@ -0,0 +1,38 @@
+
+    resource_id = event["resourceId"].downcase.split("/")
+    return nil if resource_id.length < 9
+    join(
```
I don't really like this method being separated out...would be cleaner just to have it directly in the method.

Fryguy commented on app/models/manageiq/providers/azure/event_catcher_mixin.rb:
```diff
@@ -0,0 +1,38 @@
+    join(
+      resource_id[2],
+      resource_id[4],
+      resource_id[6] + "\/" + resource_id[7],
```
Prefer interpolation here over `+`

Fryguy commented on app/models/manageiq/providers/azure/event_catcher_mixin.rb:
```diff
@@ -0,0 +1,38 @@
+    # /subscriptions/123456789-a1234-12b4-1234-5cd67890312/resourceGroups/
+    # rg_name/providers/Microsoft.Compute/virtualMachines/vm_name
+
+    resource_id = event["resourceId"].downcase.split("/")
```
Would it be possible to name the parts instead of using numbered locations?  That is change this to something like

```ruby
_, _, vm_ref, _, resource_group_name, _, something, something_else, *_rest = event["resourceId"].downcase.split("/")
```

The names are terrible, but you get the idea :smile:

Fryguy commented on db/migrate/data/20160127210624_convert_configurations_to_settings_changes/event_handling.tmpl.yml:
```diff
@@ -130,6 +130,7 @@ event_groups:
     - orchestration.autoscaling.error
     - servergroup.update
     - servergroup.addmemeber
+    - virtualMachines_write_EndRequest
```
These changes should not be in this file...they should be in settings.yml

Fryguy commented on spec/models/manageiq/providers/azure/cloud_manager/event_catcher/runner_spec.rb:
```diff
@@ -0,0 +1,39 @@
+      allow_any_instance_of(MiqWorker::Runner).to receive(:worker_initialization)
+    end
+
+    describe "filter events" do
```
This should describe the method-under-test...so this should say `describe "#filtered?" do`

Fryguy commented on spec/models/manageiq/providers/azure/cloud_manager/event_catcher/runner_spec.rb:
```diff
@@ -0,0 +1,39 @@
+    end
+
+    describe "filter events" do
+      it "filter" do
```
and this should describe something about what you are filtering, so should say something like `it "with a filtered event" do`

Fryguy commented on spec/models/manageiq/providers/azure/cloud_manager/event_catcher/runner_spec.rb:
```diff
@@ -0,0 +1,39 @@
+        expect(catcher.filtered?(event)).to be true
+      end
+
+      it "do not filter" do
```
...and this should be something like `it "with an event to keep" do`

Fryguy commented on spec/models/manageiq/providers/azure/cloud_manager/event_catcher/runner_spec.rb:
```diff
@@ -0,0 +1,39 @@
+      end
+    end
+
+    describe "parse properties" do
```
This should be broken up into two separate describes...one for each method-under-test.  And since you only have one test inside...the `it`s can just be by themselves without the describe...

```ruby
it "#parse_event_type" do
  ...
end

it "#parse_vm_ref" do
  ...
end
```

Fryguy commented on app/models/manageiq/providers/azure/cloud_manager/event_catcher/runner.rb:
```diff
@@ -0,0 +1,57 @@
+    event_monitor_handle.start
+    event_monitor_handle.each_batch do |events|
+      event_monitor_running
+      if events && !events.empty?
```
Prefer `if events.present?` over `if events && !events.empty?`

Fryguy replied:
On that note, why would the handle even emit an empty set of events (or nil for that matter)....more a question for @blomquisg if this is a normal pattern

durandom replied:
@Fryguy @blomquisg indeed this check can be removed, because in this case `handle.each_batch` only yields arrays of events.

Fryguy commented on app/models/manageiq/providers/azure/cloud_manager/event_catcher/runner.rb:
```diff
@@ -0,0 +1,57 @@
+    else
+      _log.info "#{log_prefix} Caught event #{parse_event_type(event)} for #{event["resourceId"]}"
+      EmsEvent.add_queue('add_azure', @cfg[:ems_id], event)
+
```
Unneeded blank line.

Fryguy commented:
@bronaghs Looks really good!  Most of my comments were style and structure, so relatively minor.

bronaghs commented:
thanks @Fryguy !

bronaghs commented:
@blomquisg @Fryguy @durandom 
I addressed your feedback and kept the changes in a separate commit to make it easy to identify. I don't think i rejected any of your feedback. Thanks for taking the time to review it.

lfu commented:
The event switchboard modeling looks good.

durandom commented:
looks good to me, nice job @bronaghs 👍 

only as a side note for future specs, if you run this...
```
❯ be rspec --format documentation spec/models/manageiq/providers/azure/cloud_manager/event_catcher/runner_spec.rb

ManageIQ::Providers::Azure::CloudManager::EventCatcher::Runner
  parsing properties
    parse properties
      event type
      vm ref
```
...it should read like documentation. That also helps reading the specs in order to get a clue on how the code is supposed to work and to fix broken specs. It also helps structuring the specs in a nice clean way.

blomquisg commented:
What the heck @durandom?!  Holding out on the cool tricks! :)
